src/target/riscv/riscv.c

```diff
@@ -477,203 +556,227 @@ static int maybe_add_trigger_t1(struct target *target,
-				PRIx64 " to tdata1 it contains 0x%" PRIx64,
-				tdata1, tdata1_rb);
-		riscv_set_register(target, GDB_REGNO_TDATA1, 0);
-		return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
-	}
-
-	riscv_set_register(target, GDB_REGNO_TDATA2, trigger->address);
-
+	tdata1 = set_field(tdata1, bpcontrol_bpmatch, 0); /* exact match */
+	tdata2 = trigger->address;
+	ret = set_trigger(target, idx, tdata1, tdata2);
+	if (ret != ERROR_OK)
+		return ret;
+	r->trigger_unique_id[idx] = trigger->unique_id;
 	return ERROR_OK;
 }
 
-static int maybe_add_trigger_t2(struct target *target,
-		struct trigger *trigger, uint64_t tdata1)
+static int maybe_add_trigger_t2(struct target *target, struct trigger *trigger)
 {
-	RISCV_INFO(r);
+	int idx, ret;
+	riscv_reg_t tdata1, tdata2;
 
-	/* tselect is already set */
-	if (tdata1 & (CSR_MCONTROL_EXECUTE | CSR_MCONTROL_STORE | CSR_MCONTROL_LOAD)) {
-		/* Trigger is already in use, presumably by user code. */
-		return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
-	}
-
-	/* address/data match trigger */
-	tdata1 |= CSR_MCONTROL_DMODE(riscv_xlen(target));
-	tdata1 = set_field(tdata1, CSR_MCONTROL_ACTION,
-			CSR_MCONTROL_ACTION_DEBUG_MODE);
-	tdata1 = set_field(tdata1, CSR_MCONTROL_MATCH, CSR_MCONTROL_MATCH_EQUAL);
-	tdata1 |= CSR_MCONTROL_M;
-	if (r->misa & (1 << ('S' - 'A')))
-		tdata1 |= CSR_MCONTROL_S;
-	if (r->misa & (1 << ('U' - 'A')))
-		tdata1 |= CSR_MCONTROL_U;
-
-	if (trigger->execute)
-		tdata1 |= CSR_MCONTROL_EXECUTE;
-	if (trigger->read)
-		tdata1 |= CSR_MCONTROL_LOAD;
-	if (trigger->write)
-		tdata1 |= CSR_MCONTROL_STORE;
-
-	riscv_set_register(target, GDB_REGNO_TDATA1, tdata1);
-
-	uint64_t tdata1_rb;
-	int result = riscv_get_register(target, &tdata1_rb, GDB_REGNO_TDATA1);
-	if (result != ERROR_OK)
-		return result;
-	LOG_DEBUG("tdata1=0x%" PRIx64, tdata1_rb);
+	RISCV_INFO(r);
 
-	if (tdata1 != tdata1_rb) {
-		LOG_DEBUG("Trigger doesn't support what we need; After writing 0x%"
-				PRIx64 " to tdata1 it contains 0x%" PRIx64,
-				tdata1, tdata1_rb);
-		riscv_set_register(target, GDB_REGNO_TDATA1, 0);
-		return ERROR_TARGET_RESOURCE_NOT_AVAILABLE;
-	}
+	tdata1 = 0;
+	tdata1 = set_field(tdata1, CSR_MCONTROL_TYPE(riscv_xlen(target)), 2);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_DMODE(riscv_xlen(target)), 1);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_ACTION, CSR_MCONTROL_ACTION_DEBUG_MODE);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_M, 1);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_S, !!(r->misa & (1 << ('S' - 'A'))));
+	tdata1 = set_field(tdata1, CSR_MCONTROL_U, !!(r->misa & (1 << ('U' - 'A'))));
+	tdata1 = set_field(tdata1, CSR_MCONTROL_EXECUTE, trigger->execute);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_LOAD, trigger->read);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_STORE, trigger->write);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_SIZELO, CSR_MCONTROL_SIZELO_ANY & 3);
+	tdata1 = set_field(tdata1, CSR_MCONTROL_SIZEHI, (CSR_MCONTROL_SIZELO_ANY >> 2) & 3);
+
+	if (trigger->execute || trigger->length == 1)
```

Comment:
Why do execute triggers have to use MATCH_EQUAL? I could imagine somebody wanting to trigger when an instruction anywhere on the stack is executed, for instance. Maybe gdb/OpenOCD has no way to set a breakpoint like that?

Reply:
> OpenOCD can set breakpoints on an address range:
> 
>     * https://openocd.org/doc/html/General-Commands.html#index-breakpoint
> 
> 
> I don't think that gdb can:
> 
>     * https://ftp.gnu.org/old-gnu/Manuals/gdb/html_node/gdb_28.html#SEC29

gdb should get the length information by analyzing the debugging information
gdb and openocd communicate via GDB Remote Serial Protocol

‘z0,addr,kind’
‘z1,addr,kind’
‘z2,addr,kind’
‘z3,addr,kind’
‘z4,addr,kind’

‘Z0,addr,kind’
‘Z1,addr,kind’
‘Z2,addr,kind’
‘Z3,addr,kind’
‘Z4,addr,kind’

The default kind is the length
https://github.com/bminor/binutils-gdb/blob/master/gdb/remote.c#L10665-L10700